ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -48,97 +52,119 @@ private void initializeProtoArrayForkChoice() {
-                        beaconState.getSlot(),
-                        beaconState.getLatest_block_header().hash_tree_root())));
-
-    if (!result.isSuccessful()) {
-      return result;
-    }
-
-    transaction.commit().join();
-    result
-        .getBlockProcessingRecord()
-        .ifPresent(record -> forkChoiceStrategy.onBlock(block.getMessage(), record.getPostState()));
-
-    return result;
+    return withLock(
+        () -> {
+          final ForkChoiceStrategy forkChoiceStrategy = getForkChoiceStrategy();
+          final StoreTransaction transaction = recentChainData.startStoreTransaction();
+          final BlockImportResult result =
+              on_block(
+                  transaction,
+                  block,
+                  preState,
+                  stateTransition,
+                  forkChoiceStrategy,
+                  beaconState ->
+                      transaction.putStateRoot(
+                          beaconState.hash_tree_root(),
+                          new SlotAndBlockRoot(
+                              beaconState.getSlot(),
+                              beaconState.getLatest_block_header().hash_tree_root())));
+
+          if (!result.isSuccessful()) {
+            return SafeFuture.completedFuture(result);
+          }
+          return transaction
+              .commit()
+              .thenRun(
+                  () ->
+                      result
+                          .getBlockProcessingRecord()
+                          .ifPresent(
+                              record ->
+                                  forkChoiceStrategy.onBlock(
+                                      block.getMessage(), record.getPostState())))
+              .thenApply(__ -> result);
+        });
   }
 
   public SafeFuture<AttestationProcessingResult> onAttestation(
       final ValidateableAttestation attestation) {
     return recentChainData
         .retrieveCheckpointState(attestation.getData().getTarget())
-        .thenApply(
-            targetState -> {
-              StoreTransaction transaction = recentChainData.startStoreTransaction();
-              final AttestationProcessingResult result =
-                  on_attestation(transaction, attestation, targetState, getForkChoiceStrategy());
-              if (result.isSuccessful()) {
-                transaction.commit(() -> {}, "Failed to persist attestation result");
-              }
-              return result;
-            });
+        .thenCompose(
+            targetState ->
+                withLock(
+                    () -> {
+                      final StoreTransaction transaction = recentChainData.startStoreTransaction();
+                      final AttestationProcessingResult result =
+                          on_attestation(
+                              transaction, attestation, targetState, getForkChoiceStrategy());
+                      return transaction.commit().thenApply(__ -> result);
```

Comment:
Seems like we don't check if the attestation processing was successful before committing anymore. Is this on purpose?

Reply:
No. Good spot we need to fix that.